File: Day01/Day01.java
import java.util.ArrayList;
import java.util.Scanner;
import java.util.HashMap;

public class Day01 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        boolean part1mode = false; // true = Part 1; false = Part 2

        // parse input
        ArrayList<Integer> l = new ArrayList<Integer>();
        ArrayList<Integer> r = new ArrayList<Integer>();
        String s = "";
        String[] numbers;
        while(sc.hasNext()) {
            s = sc.nextLine().trim().replaceAll(" +", " ");
            numbers = s.split(" ");
            l.add(Integer.parseInt(numbers[0]));
            r.add(Integer.parseInt(numbers[1]));
        }

        if(part1mode) {
            // PART 1
            long sum = 0;
            l.sort(null);
            r.sort(null);
            for(int i = 0; i < l.size(); i++) {
                sum += Math.abs(l.get(i) - r.get(i));
            }
            System.out.println(sum);
        } else {
            // PART 2
            long sum = 0;
            HashMap<Integer, Integer> counts = new HashMap<Integer, Integer>();
            for(Integer i: r) {
                counts.put(i, (counts.containsKey(i) ? counts.get(i) : 0) + 1);
            }
            for(Integer i: l) {
                sum += i * (counts.containsKey(i) ? counts.get(i) : 0);
            }
            System.out.println(sum);
        }

        sc.close();
    }
}

File: Day01/Day01.py
#!/usr/bin/env python
import sys

# parse input
l, r, result = [], [], 0
part1mode = False

for line in sys.stdin:
    s = line.split()
    l.append(int(s[0]))
    r.append(int(s[1]))

if(part1mode):
    # part 1
    l.sort()
    r.sort()

    for il, ir in zip(l, r):
        result += abs(il - ir)

else:
# part 2
    lookup = {}

    for i in r:
        lookup.update({i: lookup[i] + 1 if i in lookup else 1})
    
    for i in l:
        result += i * lookup[i] if i in lookup else 0
        
print(result)










